sdk/src/account_overrides.rs

```diff
@@ -0,0 +1,138 @@
+use {
+    serde::{
+        de::{self, Visitor},
+        ser::SerializeMap,
+        Deserialize, Serialize,
+    },
+    solana_sdk::{
+        account::{Account, AccountSharedData},
+        pubkey::{ParsePubkeyError, Pubkey},
+        stake_history::Epoch,
+        sysvar,
+    },
+    std::collections::HashMap,
+};
+
+/// Encapsulates overridden accounts, typically used for transaction simulations
+#[derive(Default, PartialEq, Eq, Debug, Clone)]
+pub struct AccountOverrides {
+    accounts: HashMap<Pubkey, AccountSharedData>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct AccountDef {
+    lamports: u64,
+    data: Vec<u8>,
+    owner: String,
+    executable: bool,
+    rent_epoch: Epoch,
+}
+
+impl From<AccountSharedData> for AccountDef {
+    fn from(account: AccountSharedData) -> Self {
+        let Account {
+            lamports,
+            data,
+            owner,
+            executable,
+            rent_epoch,
+        } = account.into();
+
+        AccountDef {
+            lamports,
+            data,
+            owner: format!("{}", owner),
+            executable,
+            rent_epoch,
+        }
+    }
+}
+
+impl TryFrom<AccountDef> for AccountSharedData {
+    type Error = ParsePubkeyError;
+
+    fn try_from(def: AccountDef) -> Result<Self, Self::Error> {
+        Ok(Account {
+            lamports: def.lamports,
+            data: def.data,
+            owner: def.owner.parse()?,
+            executable: def.executable,
+            rent_epoch: def.rent_epoch,
+        }
+        .into())
+    }
+}
+
+impl Serialize for AccountOverrides {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(self.accounts.len()))?;
+        for (pubkey, account) in self.accounts.iter() {
+            map.serialize_entry(&format!("{}", pubkey), &AccountDef::from(account.clone()))?;
+        }
+        map.end()
+    }
+}
```

Comment:
Why there are manual implementations for `(de)serialize`?

Also fyi, hashmap serializations are not stable:
https://github.com/bincode-org/bincode/issues/230

Reply:
I'm not implementing Serialize/Deserialize because of bincode, but so that it can be sent as a JSON in the `simulateTransaction` parameter. It's a manual implementation because I want the hashes to be base58-encoded only for this specific struct, and manually implementing that was the easiest approach I came up with.

Reply:
> I'm not implementing Serialize/Deserialize because of bincode

That is not bincode specific. serde does not have stable hashmap serializations;
I don't know if it matters here or not, though.

For (de)serialization implementation, you wouldn't need any of `AccountOverridesVisitor`, `serialize_map`, etc.
You can just convert `AccountOverrides` into or from `HashMap<String, AccountDef>`, and call `(de)serialize` on the `HashMap`.

Reply:
@behzadnouri I made it that way so that it is deserialized directly from the JSON and we can get more meaningful errors during the deserialization step, though. I do believe this is a better approach than delaying the conversion of String to Pubkey, which would make it a bit more difficult to report the error back to the deserializer.